uber/automated_emails.py

```diff
@@ -46,46 +64,235 @@ def computed_subject(self, x):
+
+    @property
+    def approved(self):
+        """
+        Check if this email category has been approved by the admins to send automated emails.
+
+        :return: True if we are approved to send this email, or don't need approval. False otherwise
+
+        Side effect: If running as part of the automated email daemon code, and we aren't approved, log the count of
+        emails that would have been sent so we can report it via the UI later.
+        """
+
+        approved_to_send = not self.needs_approval or self.ident in c.EMAIL_APPROVED_IDENTS
+
+        if not approved_to_send:
+            # log statistics about how many emails would have been sent if we had approval.
+            # if running as part of a daemon, this will record the data.
+            SendAllAutomatedEmailsJob.log_unset_because_unapproved(self)
+
+        return approved_to_send
+
+    def render(self, model_instance):
+        model = getattr(model_instance, 'email_model_name', model_instance.__class__.__name__.lower())
+        return render('emails/' + self.template, dict({model: model_instance}, **self.extra_data))
 
-    def render(self, x):
-        model = getattr(x, 'email_model_name', x.__class__.__name__.lower())
-        return render('emails/' + self.template, dict({model: x}, **self.extra_data))
+    def really_send(self, model_instance):
+        """
+        Actually send an email to a particular model instance (i.e. a particular attendee).
+
+        Doesn't perform any kind of checks at all if we should be sending this, just immediately sends the email
+        no matter what.
 
-    def send(self, x, raise_errors=True):
+        NOTE: use send_if_possible() instead of calling this method unless you 100% know what you're doing.
+        """
         try:
             subject = self.computed_subject(x)
             format = 'text' if self.template.endswith('.txt') else 'html'
-            send_email(self.sender, x.email, subject, self.render(x), format, model=x, cc=self.cc, ident=self.ident)
+            send_email(self.sender, model_instance.email, subject,
+                       self.render(model_instance), format,
+                       model=model_instance, cc=self.cc, ident=self.ident)
         except:
-            log.error('error sending {!r} email to {}', self.subject, x.email, exc_info=True)
-            if raise_errors:
-                raise
+            log.error('error sending {!r} email to {}', self.subject, model_instance.email, exc_info=True)
+            raise
+
+    @property
+    def when_txt(self):
+        """
+        Return a textual description of when the date filters are active for this email category.
+        """
+
+        return '\n'.join([filter.active_when for filter in self.when])
+
+
+class SendAllAutomatedEmailsJob:
+
+    # save information about the last time the daemon ran so that we can display stats on things like
+    # unapproved emails/etc
+    last_result = dict()
+
+    run_lock = threading.Lock()
```

Comment:
I hear you on that.  Though in this particular scenario it looked like the differences between them weren't really that different for this particular setup.